feat: display tracking information about unused variables

gentzeng opened the pull request:
Unused variables are grouped per containig files and displayed with line
and column information. Files are displayed relatively to working
directory

gentzeng commented:
Sorry for the gratuitous pull requests before.

XhmikosR commented on cli.js:
```diff
@@ -27,15 +27,21 @@ function main(args) {
+        unusedVarsInfo.unused.forEach(unusedVarInfo => {
+            if (currentFile !== unusedVarInfo.file) {
+                currentFile = unusedVarInfo.file;
+                console.log('\n\u001B[4m' + currentFile + '\u001B[0m');
```
Not a huge fan of these characters. See if we can achieve the same with chalk

gentzeng replied:
Fixed

XhmikosR commented on src/app.js:
```diff
@@ -0,0 +1,10 @@
+const express = require('express');
```
Unsure why this was added.

gentzeng replied:
Happens ;)

XhmikosR commented on index.js:
```diff
@@ -10,6 +10,84 @@ const defaultOptions = {
+
+function getVariableLinePerContainingFile(unusedVarsInfo, varLineInfoInAllFiles) {
+    for (
+        let i = 0, j = 0, linesChecked = 0,
```
This is definitely not really readable

gentzeng replied:
I guess you mean the content of this function

gentzeng replied:
I admit, it's was not readable. Refactored.

danny007in commented on index.js:
```diff
@@ -10,6 +10,107 @@ const defaultOptions = {
     ignore: []
 };
 
+function getLineCountOfSassFile(sassFileString) {
```
change 
```js
function getLineCountOfSassFile(sassFileString) {
```
To
```js
// To get the number of lines in single sass files
function totalLines(file) {
```

And do similar thing for all u edited, so that can easily readable. Add Comment above method rather than lengthy method name

gentzeng replied:
Done

danny007in replied:
ok

gentzeng replied:
Actually wanted it to be close to output form stylelint.

For example like this:

![relativePath](https://user-images.githubusercontent.com/60478031/106204934-06c22780-61be-11eb-945a-afecbba00304.png)

Furthermore, I wanted the path output to be consistent with sonarqube general issues format.
![sonar](https://user-images.githubusercontent.com/60478031/106205009-1f324200-61be-11eb-9700-9aa67fe7a7e0.png)

The latter picture is the result of another feature, I will request to get pulled after this one, since it build upon it. This second feature allows to import fusv results to sonarqube. Hence, the emphasis on displaying paths consistent with respect to sonarqube and related linters like stylelint.

danny007in commented on index.js:
```diff
@@ -10,6 +10,122 @@ const defaultOptions = {
+
+    sassFiles.forEach(file => {
+        const sassStr = fs.readFileSync(file, 'utf8');
+        const fileRelativePath = path.relative(currentDirectory, file);
```
get path from root,
like `C:/xxxxx`

gentzeng replied:
Actually, I wanted it to be close to the output format of, e.g., stylelint.

For example like this:

![relativePath](https://user-images.githubusercontent.com/60478031/106204934-06c22780-61be-11eb-945a-afecbba00304.png)

Furthermore, I wanted the path output to be consistent with sonarqube general issues format.
![sonar](https://user-images.githubusercontent.com/60478031/106205009-1f324200-61be-11eb-9700-9aa67fe7a7e0.png)

The latter picture is the result of another feature, which I will request to get pulled right after this one, since it builds upon it. This second feature allows to import fusv results to sonarqube. Hence, the emphasis on displaying paths consistent with respect to sonarqube and related linters like stylelint.

danny007in replied:
ok

danny007in commented on tests/subtest/_variables2.scss:
```diff
@@ -0,0 +1,5 @@
+// unused variables
```
change file name to `_another_variables`

danny007in commented on tests/subtest/_variables2.scss:
```diff
@@ -0,0 +1,5 @@
+// unused variables
```
This file not required

gentzeng replied:
Fixed

danny007in replied:
If you have fixed then click on Resolve Button 😄 , only if ur 100% sure

danny007in commented on index.js:
```diff
@@ -10,6 +10,121 @@ const defaultOptions = {
+    return singleFilesInfo;
+}
+
+function removeJekyllComments(sassFileString) {
```
make reusable function only if its required more than once

danny007in commented on cli.js:
```diff
@@ -27,15 +27,24 @@ function main(args) {
 
         // eslint-disable-next-line unicorn/no-array-callback-reference
-        const unusedVars = fusv.find(dir, { ignore });
+        const unusedVarsInfo = fusv.find(dir, { ignore });
```
no need change variable name

danny007in commented on lib/parse-variable.js:
```diff
@@ -11,15 +11,28 @@ function parseNodes(nodes, variables, ignoreList) {
     }
 }
 
+function isUnignoredVariable(node, ignoreList) {
```
making more complicated by creating unwanted reusable method.

danny007in replied:
if method is not reused more than once, then not required

danny007in commented:
@gentzeng before push run `npm run test` first

danny007in commented on index.js:
```diff
@@ -10,6 +10,112 @@ const defaultOptions = {
+    return lineCount;
+}
+
+function concatSassFiles(sassFiles) {
```
if method is not reused more than once, then not required

danny007in commented on index.js:
```diff
@@ -10,6 +10,112 @@ const defaultOptions = {
+//      path relatively to given sass file directory
+//      linesCumulated: sum of all files before and including
+//      the currently examined file
+function getFilesInfo(sassFiles) {
```
if method is not reused more than once, then not required

danny007in commented on index.js:
```diff
@@ -10,6 +10,112 @@ const defaultOptions = {
+
+// Filter the unused variables from variables by checking
+// the number of occurences in sassFilesString
+function filterUnusedVars(variables, sassFilesString) {
```
if method is not reused more than once, then not required

danny007in commented on index.js:
```diff
@@ -10,6 +10,112 @@ const defaultOptions = {
+    return unusedVarsInfo;
+}
+
+function calcLineOfVar(varInfo, singleFilesInfo) {
```
if method is not reused more than once, then not required

gentzeng replied:
This function is important for the functionality of the wrapping loop. Putting it back would lead to an already resolved conversation.

danny007in replied:
ok

danny007in replied:
then make it private method `_calcLineOfVar`

danny007in commented on index.js:
```diff
@@ -10,6 +10,112 @@ const defaultOptions = {
+    return varInfo;
+}
+
+function calcVariableLines(unusedVarsInfo, singleFilesInfo) {
```
if method is not reused more than once, then not required

danny007in commented on index.js:
```diff
@@ -10,6 +10,112 @@ const defaultOptions = {
     ignore: []
 };
 
+function totalLines(sassFileString) {
```
if method is not reused more than once, then not required

danny007in commented on index.js:
```diff
@@ -10,6 +10,112 @@ const defaultOptions = {
+    // Since finding unused variables work on a "catFile",
+    // i.e., a concatenation of all files, we have to calculate lines per file
+    // from variable lines in the catFile
+    const varLineInAllFiles = varInfo.lineInAllFiles;
```
avoid this declared `varLineInAllFiles`

danny007in commented on index.js:
```diff
@@ -28,29 +134,24 @@ function findUnusedVars(strDir, opts) {
+    unusedVarsInfo = calcVariableLines(unusedVarsInfo, singleFilesInfo);
 
     return {
-        unused: unusedVars,
```
don not rename simply plz, 😃 
because while reviewing code its get difficult 🕵🏼

danny007in commented on lib/parse-variable.js:
```diff
@@ -11,15 +11,28 @@ function parseNodes(nodes, variables, ignoreList) {
-    if (node instanceof Declaration && node.prop.startsWith('$') && !ignoreList.includes(node.prop) && fusvEnabled) {
-        result.push(node.prop);
+    if (isUnignoredVariable(node, ignoreList)) {
+        const nodeInfo = {
```
directly add in 
```js
result.push({           
   lineInAllFiles: node.source.start.line,
   column: node.source.start.column,
   name: node.prop
})
```

danny007in replied:
because `nodeInfo` is not used more than one, and also the objects are not lenthy

danny007in commented on index.js:
```diff
@@ -28,29 +134,24 @@ function findUnusedVars(strDir, opts) {
         sassFilesString = sassFilesString.replace(/---/g, '');
     }
 
-    const variables = parse(sassFilesString, options.ignore);
```
avoid rename

danny007in commented on cli.js:
```diff
@@ -31,8 +31,17 @@ function main(args) {
-        unusedVars.unused.forEach(unusedVar => {
-            console.log(`Variable ${chalk.bold(unusedVar)} is not being used!`);
+        let currentFile = '';
+        unusedVars.unused.forEach(unusedVarInfo => {
```
avoid rename `unusedVarInfo `

danny007in commented on cli.js:
```diff
@@ -31,8 +31,17 @@ function main(args) {
+                console.log(chalk.underline(currentFile));
+            }
+
+            console.log(
```
make 2 logs

gentzeng replied:
Not sure, what you mean. the log in line 41 cannot be in to log commands since then an unwanted newline would be displayed. I wrote is so it would not be to far right.

danny007in replied:
ok

danny007in commented on lib/parse-variable.js:
```diff
@@ -11,15 +11,27 @@ function parseNodes(nodes, variables, ignoreList) {
         return;
     }
 
-    if (node instanceof Declaration && node.prop.startsWith('$') && !ignoreList.includes(node.prop) && fusvEnabled) {
```
no need of change here.
and plz reply before mark as resolved

gentzeng replied:
Sorry. thought, I changed it already

gentzeng replied:
done

danny007in replied:
ok

danny007in commented on index.js:
```diff
@@ -41,16 +77,51 @@ function findUnusedVars(strDir, opts) {
-        const re = new RegExp(`(${escapeRegex(variable)})\\b(?!-)`, 'g');
+    // Filter the unused variables from variables by checking
+    // the number of occurences in sassFilesString
+    const unusedVars = variables.filter(variableInfo => {
```
avoid rename `variableInfo`

gentzeng replied:
done

danny007in commented on cli.js:
```diff
@@ -31,8 +31,17 @@ function main(args) {
+                console.log(`\n${chalk.underline(currentFile)}`);
+            }
+
+            console.log(
+                ` ${unusedVar.lineInOwnFile}:${unusedVar.column}\t` +
+                `Variable ${chalk.bold(unusedVar.name)} is not being used!`
+            );
```
add under `            if (currentFile !== unusedVar.file) {`

danny007in replied:
probably good idea

gentzeng replied:
As far as I can tell, it already is under `Sorry. thought, I changed it already`

danny007in replied:
no its not

gentzeng replied:
If you mean, if it is not with in the block starting with `if (currentFile !== unusedVar.file) {`:
I put it below the block so the file is not displayed new per variable, but only once. (Again with emphasis on consistence with linters like stylelint)

![variablegroups](https://user-images.githubusercontent.com/60478031/106213617-5d832d80-61cd-11eb-96db-746ce9c11704.png)

danny007in replied:
ok

danny007in commented on index.js:
```diff
@@ -28,10 +63,11 @@ function findUnusedVars(strDir, opts) {
-    let sassFilesString = sassFiles.reduce((sassStr, file) => {
-        sassStr += fs.readFileSync(file, 'utf8');
-        return sassStr;
+    let sassFilesString = sassFiles.reduce((sassStrCat, file) => {
```
avoid rename `sassStrCat`

gentzeng replied:
done

danny007in commented on index.js:
```diff
@@ -28,9 +63,9 @@ function findUnusedVars(strDir, opts) {
     // Array of all Sass files
     const sassFiles = glob.sync(path.join(dir, '**/*.scss'));
 
-    // String of all Sass files' content
```
y removed

danny007in commented on index.js:
```diff
@@ -28,9 +63,9 @@ function findUnusedVars(strDir, opts) {
-    // String of all Sass files' content
     let sassFilesString = sassFiles.reduce((sassStr, file) => {
         sassStr += fs.readFileSync(file, 'utf8');
+
```
why new line simply

gentzeng replied:
Because the xo linter is telling to put a line right before returns of functions. It does not do taht within callbacks like here, but a added this line for the sake of consistency: Always a an empty line before return. Your thoughts?

danny007in replied:
cool, relax, take a deep breath and  😄 .
u could just reply shortly as 'lint error'

gentzeng replied:
Haha, ok^^

danny007in commented on index.js:
```diff
@@ -41,16 +77,50 @@ function findUnusedVars(strDir, opts) {
     return {
         unused: unusedVars,
-        total: variables.length
+        total: unusedVars.length
```
variables length also required

danny007in commented on index.js:
```diff
@@ -41,16 +77,50 @@ function findUnusedVars(strDir, opts) {
         return sassFilesString.match(re).length === 1;
     });
 
+    // Get per file:
+    //      path relatively to given sass file directory
+    //      linesCumulated: sum of all files before and including
+    //      the currently examined file
+    const singleFilesInfo = [];
+    const currentDirectory = process.cwd();
+
+    let linesCumulated = 0;
+
+    sassFiles.forEach(file => {
+        const sassStr = fs.readFileSync(file, 'utf8');
+        const fileRelativePath = path.relative(currentDirectory, file);
+
+        let totalLines = 0;
+        for (let i = 0, n = sassStr.length; i < n; ++i) {
+            if (sassStr[i] === '\n') {
+                ++totalLines;
+            }
+        }
+
+        linesCumulated += totalLines;
+
+        singleFilesInfo.push({
+            file: fileRelativePath,
+            linesCumulated
+        });
+    }, '');
+
+    // calculate line information per variable
+    for (let varInfo of unusedVars) {
+        varInfo = calcLineOfVar(varInfo, singleFilesInfo);
+    }
+
```
unnecessary code

danny007in commented on index.js:
```diff
@@ -10,6 +10,41 @@ const defaultOptions = {
     ignore: []
 };
 
+function calcLineOfVar(varInfo, singleFilesInfo) {
```
u may get by here https://github.com/XhmikosR/find-unused-sass-variables/blob/3590386e9d669ae1c906058afea1b27300900b61/lib/parse-variable.js#L8

`nodes`

danny007in commented on index.js:
```diff
@@ -41,16 +77,50 @@ function findUnusedVars(strDir, opts) {
+
     return {
         unused: unusedVars,
-        total: variables.length
```
this line is required

gentzeng replied:
fixed

danny007in commented on index.js:
```diff
@@ -41,16 +77,50 @@ function findUnusedVars(strDir, opts) {
     return {
         unused: unusedVars,
-        total: variables.length
+        total: unusedVars.length
```
change to
```js
        totalUnusedVars: unusedVars.length
```

gentzeng replied:
done

gentzeng commented:
I just realized I introduced some kind of a breaking change. Hence, I delay this pull request until I fixed that.

danny007in commented:
It's simple to do,  I have can another method

gentzeng commented:
I fixed this breaking change.
find() now again returns `total` and `unused` to be consistent with the given API. However, I added totalUnused (as you recommended) and further `unsuedOrigin`, whichs is containing the `file`, `line` and `column` of a variable as return values of find.

gentzeng commented:
I don't seem to be able to reopen this pull request. Would you please? Thanks in advance :)
